src/hotspot/share/gc/epsilon/epsilonHeap.cpp

```diff
@@ -110,26 +110,36 @@ HeapWord* EpsilonHeap::allocate_work(size_t size) {
 
   while (res == NULL) {
     // Allocation failed, attempt expansion, and retry:
-    MutexLocker ml(Heap_lock);
-
-    size_t space_left = max_capacity() - capacity();
-    size_t want_space = MAX2(size, EpsilonMinHeapExpand);
-
-    if (want_space < space_left) {
-      // Enough space to expand in bulk:
-      bool expand = _virtual_space.expand_by(want_space);
-      assert(expand, "Should be able to expand");
-    } else if (size < space_left) {
-      // No space to expand in bulk, and this allocation is still possible,
-      // take all the remaining space:
-      bool expand = _virtual_space.expand_by(space_left);
-      assert(expand, "Should be able to expand");
-    } else {
-      // No space left:
-      return NULL;
-    }
+    {
```

Comment:
I see what you are trying to do, and it makes sense. I believe this form would be cleaner:

```
HeapWord* res = NULL;
while (true) {
  // Try to allocate, assume space is available
  res = par_allocate(size);
  if (res != NULL) {
    break;
  }

  MutexLocker ml(Heap_Lock);

  // Try to allocate under the lock, assume another thread was able to expand
  res = par_allocate(size);
  if (res != NULL) {
    break;
  }

  // Expand and loop back if space is available
  size_t space_left = max_capacity() - capacity();
  size_t want_space = MAX2(size, EpsilonMinHeapExpand);
   ...
}
```

Reply:
Yes, the new form is cleaner very. And I think it would be a little cleaner if wrap the lock scope in curly braces. like this:

```c++
HeapWord* res = NULL;
while (true) {
  // Try to allocate, assume space is available
  res = par_allocate(size);
  if (res != NULL) {
    break;
  }

  {
    MutexLocker ml(Heap_Lock);
  
    // Try to allocate under the lock, assume another thread was able to expand
    res = par_allocate(size);
    if (res != NULL) {
      break;
    }
  
    // Expand and loop back if space is available
    size_t space_left = max_capacity() - capacity();
    size_t want_space = MAX2(size, EpsilonMinHeapExpand);
     ...
  }
}
```